[merged] Follow device symlinks in DEVS

ianabc opened the pull request:
This patch just checks if the content of DEVS is a symlink and if it is, follows it.

I'm adding this to avoid problems I've had with devices being renamed when they are reattached in openstack (but the problem is more general).

rhatdan commented:
bot, add author to whitelist

rhatdan commented:
bot, test pull request

rhatdan commented:
LGTM

rhatdan commented:
@rhvgoyal PTAL

rhvgoyal commented on docker-storage-setup.sh:
```diff
@@ -632,6 +631,9 @@ canonicalize_block_devs() {
     # Looks like we allowed just device name (sda) as valid input. In
     # such cases /dev/$dev should be a valid block device.
+    if [ -h "$dev" ];then
+      dev=$(readlink -f $dev)
```
Why not use `readlink -e $dev` ?

rhvgoyal commented:
Can you please also add a test for making sure symlink to devices work.

ianabc commented:
I just copied one of the existing tests (004) and modified it slightly to create symlinks to the test devices. I'm doing cleanup inside the test for those symlinks

rhvgoyal commented on docker-storage-setup.sh:
```diff
@@ -632,7 +632,7 @@ canonicalize_block_devs() {
     # such cases /dev/$dev should be a valid block device.
     if [ -h "$dev" ];then
-      dev=$(readlink -f $dev)
+      dev=$(readlink -e $dev)
```
What happens if redlink fails. If destination device does not exist. Will I get a proper error message about this.

rhvgoyal replied:
Also you can modify your existing patch and do a force push. Don't have to create incremental patch for taking care of review feedback.

ianabc replied:
Im not really sure how to modify the existing patch and force push.

rhvgoyal commented on tests/009-test-follow-symlinked-devices.sh:
```diff
@@ -0,0 +1,50 @@
+  return $test_status
+}
+
+# Make sure non-absolute disk names are supported. Ex. sdb.
```
This comment is stale and will need change.

rhvgoyal commented on docker-storage-setup.sh:
```diff
@@ -627,9 +626,18 @@ check_wipe_block_dev_sig() {
+    if [ -h "$dev" ];then
+      linktarget=$(readlink -e $dev)
+      if [ $? -ne 0 ] || [ ! -b "$linktarget" ] ; then
+        Fatal "$linktarget doesn't link to a valid block device."
```
We already do block device checks on dev_abs so no need to do it again. How about something like.

if ! dest_dev=$(readlink -e $dev);then
    Fatal "readlink failed to resolve device path $dev"
fi

rhvgoyal commented on tests/009-test-follow-symlinked-devices.sh:
```diff
@@ -0,0 +1,50 @@
+  if [ $? -ne 0 ]; then
+    cleanup $vg_name "$TEST_DEVS"
+    for dev in $devs; do
+      if [ -h $dev ];then
```
We don't require this `-h` test again. We know all $dev are soft links as we have created these just now.

rhvgoyal replied:
Also it probably is good to cleanup symlinks first and then call `cleanup()`

rhvgoyal commented on tests/009-test-follow-symlinked-devices.sh:
```diff
@@ -0,0 +1,50 @@
+    if [ -h $dev ];then
+      rm $dev
+    fi 
+  done
```
This repetition of cleanup logic feels odd. Can we just define a helper function and call that function at both the places.

cleanup_soft_links()

ianabc commented:
I added cleanup_soft_links based on other routines in libtest.sh. dev isn't defined as local, but that seems to be the convention for the other subroutines in there so I just followed it.

rhvgoyal commented on docker-storage-setup.sh:
```diff
@@ -628,8 +628,16 @@ check_wipe_block_dev_sig() {
+    # If the device name is a symlink, follow it and use the target
+    if [ -h "$dev" ];then
+      if ! dest_dev=$(readlink -e $dev);then
+        Fatal "$dest_dev failed to resolve linked device path $dev"
```
When error happens $dest_dev will be empty string. Right? So it will not help using dest_dev in error message. So how about just keeping it simple as.

`Fatal "Failed to resolve symbolic link $dev"`

ianabc replied:
Yep, you're right. I've made that change and pushed it now.

rhvgoyal commented:
I do think `dev` should be local in all routines of libtest.sh. This is not intentional. I think we just could not spot it. If you like you can open another PR to clean it up. Or atleast make dev local in the function you are introducing.

rhvgoyal commented:
I tested it and automated test does not seem to be cleaning up symlink properly. If I pass in two devices, one of the symlinks seems to be left behind.

rhvgoyal commented:
These tests are assuming that device passed in TEST_DEVS is a regular device. Can we check device. And if it is not symlink, then setup a symlink for test otherwise not. That way, One can pass in symlinks in DEVS= and still be able to run automated tests.

ianabc commented:
Yes, I can take a look at that. For cleanup I guess we want to cleanup any symlinks that were created by the test, but leave any which existed before (in TEST_DEVS).

rhvgoyal commented:
@ianabc Looking at test 004, looks like it is assuming input of format "/dev/<foo-disk>". So for now, let us keep it simple. That is for tests, we will continue to pass in devices (and not symlinks). We can have another look at it later in a separate PR.

rhvgoyal commented on tests/009-test-follow-symlinked-devices.sh:
```diff
@@ -0,0 +1,48 @@
+  for dev in $TEST_DEVS; do
+    if [ ! -h $dev ]; then
+      devlink="$dev-test.$$"
+      ln -s $dev $devlink
```
Another minor nit. Can we create these soft links in some dir like /tmp. Right now they seem to be created in cwd of process and if cleanup fails for some reason, it will just leave persistent links behind.

ianabc commented:
I've reverted the change in 004 and moved the symlinks into /tmp. For the moment I haven't removed the symlink check in 009 but if it is out of place I can get rid of it.

rhvgoyal commented on tests/009-test-follow-symlinked-devices.sh:
```diff
@@ -0,0 +1,48 @@
+    test_status=0
+  fi
+
+  cleanup_soft_links $devlinks
```
double quotes around `$devlinks`

rhvgoyal commented:
LGTM

rhvgoyal commented:
@rh-atomic-bot r+
